src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -143,15 +180,31 @@ public AnalyzerConfigOptionsResult GetOptionsForSourcePath(string sourcePath)
                                 diagnosticBuilder,
                                 config.PathToFile,
                                 _diagnosticIdCache);
+                            sectionKey.Add(section);
                         }
                     }
                 }
             }
 
-            return new AnalyzerConfigOptionsResult(
-                treeOptionsBuilder.Count > 0 ? treeOptionsBuilder.ToImmutable() : SyntaxTree.EmptyDiagnosticOptions,
-                analyzerOptionsBuilder.Count > 0 ? analyzerOptionsBuilder.ToImmutable() : AnalyzerConfigOptions.EmptyDictionary,
-                diagnosticBuilder.ToImmutableAndFree());
+            // Try to avoid creating extra dictionaries if we've already seen an options result with the
+            // exact same options
+            if (!_optionsCache.TryGetValue(sectionKey, out var result))
+            {
+                result = new AnalyzerConfigOptionsResult(
+                    treeOptionsBuilder.Count > 0 ? treeOptionsBuilder.ToImmutable() : SyntaxTree.EmptyDiagnosticOptions,
+                    analyzerOptionsBuilder.Count > 0 ? analyzerOptionsBuilder.ToImmutable() : AnalyzerConfigOptions.EmptyDictionary,
+                    diagnosticBuilder.ToImmutableAndFree());
+                _optionsCache.TryAdd(sectionKey, result);
```

Comment:
Should we throwing away result and getting from the cache if TryAdd fails so that GC can free this `AnalyzerConfigOptionsResult` while it's still cheap? If not, consider adding a discard assignment to be clear about intent.

Reply:
I don't think it really matters either way, races should be rare.

Reply:
I'm not talking about races. This change is about reducing memory pressure, and this will keep an extra class around for a while. If we were to immediately remove the reference and use the one in the cache, the GC pass would be basically free as it's still in Gen 0.

Reply:
Yeah, I'm saying that it should be rare for that to happen. Most calls will hit the cache, so you may end up with one or two extra objects, but that's not what the problem is about. Note that the cache is not perfect either, there's a chance you could end up evicting dictionaries that could be shared.

Reply:
Alright, fair enough.